### PR TITLE
Refactor roots code, clean up

### PR DESCRIPTION
**Pull request by jverzani**
This (too-large) commit does many things:

* it refactors the main iterative methods. Where before an iterator approach was used, now a better separation between updating and assessing convergence is used. The approach seems to have better performance.

* this refactor allows for this proposed interface: https://discourse.julialang.org/t/a-unified-interface-for-rootfinding/698 . Currently, the entry point is called `find_roots`, though this may change. (Currently `find_roots` is exported, but not documented.) The `fzero`/`fzeros` interface is maintained.

* the tests were consolidated to tidy things up

* the `factor` function is renamed to `polyfactor`, so that `Base.Primes` is no longer a direct dependency.

* the `polyder` function is used, as the `p'` shortcut is now removed from `Polynomials`.

* the roots.md document was tidied up

* should work cleanly with v0.6

### CODE REVIEW

File: doc/roots.md [hunk 1]
```diff
@@ -129,153 +149,140 @@ x, f(x)
...
+always work.
 
+## Finding more than one zero
 
-The `roots` function degrades as there are multiplicities:
+The bracketing methods suggest a simple algorithm to recover multiple
+zeros: partition an interval into many small sub-intervals. For those
+that bracket a root find the root. This is essentially implemented with `fzeros(f,
+a, b)`. The algorithm has problems with zeros that don't cross the
+$x$-axis and zeros which bunch together. Simple usage is often succesful
```
Comment (KristofferC) [inline doc/roots.md]: "zeros that don't cross the $x$-axis" feels a bit unclear. Maybe "zeros where the function has the same sign to the left and right of the zero". Not sure if that is better, just an idea :)

File: src/Polys/multroot.jl [hunk 2]
```diff
@@ -70,7 +70,7 @@ function pejroot(p::Poly, z0::Vector, l::Vector{Int};
     a = p2a(p) #rcoeffs(monic(p))[2:end] # an_1, an_2, ..., a2, a1, a0
 
     if wts == nothing
-        wts = map(u -> min(1, 1/abs(u)), a)
+        wts = map(u -> min(1, 1/map(abs,u)), a)
```
Comment (KristofferC) [inline src/Polys/multroot.jl]: `min.(1, 1./abs.(a))`?

File: src/Polys/multroot.jl [hunk 3]
```diff
@@ -177,12 +177,12 @@ function multroot(p::Poly;
     while degree(p0) > 0
         if degree(p0) == 1
             z = roots(p0)[1]
-            _, ind = findmin(abs(zs .- z))
+            _, ind = findmin(map(abs, zs .- z))
```
Comment (KristofferC) [inline src/Polys/multroot.jl]: `abs.(zs .- z)`? same below

File: src/bracketing.jl [hunk 4]
```diff
@@ -37,73 +40,112 @@ function _middle(x::Float64, y::Float64)
...
 
-The bracket `[a,b]` must be bounded.
+abstract AbstractBisection <: UnivariateZeroMethod
+type Bisection <: AbstractBisection end
+type A42 <: AbstractBisection end 
+
+## This is a bit clunky, as we use `a42` for bisection when we don't have `Float64` values.
+## As we don't have the `A42` algorithm implemented through `find_zero`, we adjust here.
+function find_zero{M<:AbstractBisection, T<:Real}(f, x0::Vector{T}, method::M; maxevals::Int=50, verbose::Bool=false, kwargs...)
+    x = sort(map(float, x0))
```
Comment (KristofferC) [inline src/bracketing.jl]: Just `float(x0)` should be ok.

File: src/bracketing.jl [hunk 5]
```diff
@@ -37,73 +40,112 @@ function _middle(x::Float64, y::Float64)
...
+##
+## Terminates with `x1` when the bracket length of `[x0,x2]` is `<= max(xtol, xtolrel*abs(x1))` where `x1` is the midpoint .
+## The tolerances can be set to 0, in which case, the termination occurs when `nextfloat(x0) = x2`.
+## The bracket `[a,b]` must be bounded.
+
+function init_state{T <: Float64}(method::Bisection, fs, x::Vector{T}, bracket)
+    x0, x2 = sort(x[1:2])
+    isinf(x0) && (x0 = nextfloat(x0))
+    isinf(x2) && (x2 = prevfloat(x2))
+    y0, y2 = map(fs.f, [x0, x2])
```
Comment (KristofferC) [inline src/bracketing.jl]: `fs.f.([x0, x2])`?

File: src/bracketing.jl [hunk 6]
```diff
@@ -37,73 +40,112 @@ function _middle(x::Float64, y::Float64)
...
+
+    sign(y0) * sign(y2) > 0 && error("Not a bracket")
     
-    x0, y0 = a, f(a)
-    x2, y2 = b, f(b)
+    state = UnivariateZeroState(x2, x0,
+                                      y2, y0,
```
Comment (KristofferC) [inline src/bracketing.jl]: Indentation looks a bit weird here.

File: src/derivative_free.jl [hunk 7]
```diff
@@ -119,471 +32,534 @@ end
 
 ## use f[a,b] to approximate f'(x)
 function _fbracket(a, b, fa, fb)
-    out = (fb - fa) / (b - a)
+    num, den = fb-fa, b - a
+    num==0 && den == 0 && return Inf, true
```
Comment (KristofferC) [inline src/derivative_free.jl]: space after and before `==`

File: src/derivative_free.jl [hunk 8]
```diff
@@ -119,471 +32,534 @@ end
...
-        fxn1 = o.f(xn1)
-        incfn(o)
-        
-        push!(o.xn, xn1)
-        push!(o.fxn, fxn1)
-
+## order 0
+# goal: more robust to initial guess than higher order methods
+# follows roughly algorithm described http://www.hpl.hp.com/hpjournal/pdfs/IssuePDFs/1979-12.pdf, the SOLVE button from the HP-34C
+# though some modifications were made. (Much faster as older `SOLVE`, but not quite as robust.)
```
Comment (KristofferC) [inline src/derivative_free.jl]: "Much faster "THAN" older `SOLVE`"?

File: src/derivative_free.jl [hunk 9]
```diff
@@ -119,471 +32,534 @@ end
...
+    end
 
-    update = o -> begin
-        xn = o.xn[end]
-        fxn = o.fxn[end]
+    fgamma = f(gamma); incfn(o)
+    if sign(fgamma)*sign(fbeta) < 0.0
+        o.xn0, o.xn1 = gamma, beta
+        o.fxn0, o.fxn1 = fgamma, fbeta
+        opts = deepcopy(options); opts.verbose=false
```
Comment (KristofferC) [inline src/derivative_free.jl]: `deepcopy` is notoriously slow. Perhaps better to define a custom `copy` method for `UnivariateZeroOptions`.

File: src/newton.jl [hunk 10]
```diff
@@ -1,52 +1,72 @@
...
-    for x in o
-        nothing
+## extra work to allow for complex values
+function derivative_free_setup{T<:AbstractFloat}(method::Newton, fs::CallableFunction, x0::Union{T, Complex{T}};
+                                  bracket=Nullable{Vector{T}}(),
+                                  xabstol=zero(T), xreltol=zero(T),
+                                  abstol=4*eps(T), reltol=4*eps(T),
+                                  maxevals=40, maxfnevals=typemax(Int),
+                                  verbose::Bool=false)
+    x = map(float, x0)
```
Comment (KristofferC) [inline src/newton.jl]: should be enough with `float(x0)`.

File: src/newton.jl [hunk 11]
```diff
@@ -1,52 +1,72 @@
...
+
+    if isa(x, Complex)
+        bracket = Nullable{Vector{Complex{T}}}()     # bracket makes no sense for complex input, but one is expected
```
Comment (KristofferC) [inline src/newton.jl]: This looks like a type instability. `bracket` will from the arguments be a `Nullable{Vector{T}}` but will here be a `Nullable{Vector{Complex{T}}}`. However, very little work is done with it here until the next function barrier so it is probably fine.

File: README.md [hunk 12]
```diff
@@ -123,8 +124,15 @@ fzeros(x^5 -x - 1)
 fzeros(x*(x-1)*(x-2)*(x^2 + x + 1))
 ```
 
+The `factor` command will factor a polynomial or polynomial function with Integer or rational
```
Comment (jverzani) [inline README.md]: no reason. Thanks.

File: doc/roots.md [hunk 13]
```diff
@@ -1,53 +1,61 @@
...
 ```
 
 ## Bracketing
 
-For a function $f: R \rightarrow R$ a bracket is a pair $a<b$ for
-which $f(a)\cdot f(b) < 0$. That is they have different signs. If $f$
-is continuous this forces there to be a zero on the interval $[a,b]$,
+For a function $f: R \rightarrow R$ a *bracket* is a pair $ a < b $ for
+which $f(a) \cdot f(b) < 0$. That is they have different signs. If $f$
+is a continuous function this forces there to be a zero (a $c$ with
```
Comment (jverzani) [inline doc/roots.md]: good suggetion

File: doc/roots.md [hunk 14]
```diff
@@ -129,153 +149,140 @@ x, f(x)
...
-```
 
-(Though the answer is basically on track, the algorithm takes too long
-to improve itself to the very stringent range set. For problems where
-a bracket is found, this dithering won't happen.)
+For some functions, adjusting the default tolerances may be necessary
+to achieve convergence. These include `abstol` and `reltol`, which are
+used to check if `norm(f(x)) <= max(abstol, norm(x)*reltol)`;
+`xabstol`, `xreltol`, to check if `norm(x1-x0) <= max(xabstol,
+norm(x1)*xreltol)`; and `maxevals` and `maxfnevals` to limit the
```
Comment (jverzani) [inline doc/roots.md]: On counts steps in the algorithm, one function evaluations. (Some steps have 2 or more function calls.)

File: doc/roots.md [hunk 15]
```diff
@@ -129,153 +149,140 @@ x, f(x)
...
-possible, a bracketing approach. It may need many more function calls
-than the higher-order methods. These higher-order methods can be
+number of function calls. These higher-order methods can be
 susceptible to some of the usual issues found with Newton's method:
 poor initial guess, small first derivative, or large second derivative
 near the zero.
 
+----
 
 For a classic example where basically the large second derivative is
```
Comment (jverzani) [inline doc/roots.md]: Thanks

File: doc/roots.md [hunk 16]
```diff
@@ -129,153 +149,140 @@ x, f(x)
...
 the issue, we have $f(x) = x^{1/3}$:
 
 ```
 f(x) = cbrt(x)
-x = fzero(f, 1, order=8)	# all of 2, 5, 8, and 16 fail
+x = fzero(f, 1, order=2)	# all of 2, 5, 8, and 16 fail or head off to Infinity
```
Comment (jverzani) [inline doc/roots.md]: Thanks

File: doc/roots.md (hunk 1 above)
Comment (jverzani) [inline doc/roots.md]: Yes, thanks. Will address.

File: src/bracketing.jl (hunk 6 above)
Comment (jverzani) [inline src/bracketing.jl]: Thanks

File: src/derivative_free.jl (hunk 8 above)
Comment (jverzani) [inline src/derivative_free.jl]: comment to old implementation to be removed

File: src/derivative_free.jl (hunk 9 above)
Comment (jverzani) [inline src/derivative_free.jl]: Okay, I just want to turn off verbose, so can do so a cheaper way.

File: src/newton.jl (hunk 11 above)
Comment (jverzani) [inline src/newton.jl]: I'm going to leave, as this is more for pedagogical purposes.

**Comment by jverzani** [comment]
Thanks for the review @KristofferC!!

**Pull request merged**